Libraries/Opc.Ua.Server/Subscription/MonitoredItem/MonitoredItem.cs

```diff
@@ -1761,47 +1694,34 @@ protected void InitializeQueue()
                     {
                         if (m_queueSize <= 1)
                         {
-                            m_queue = null;
+                            Utils.SilentDispose(m_dataChangeQueueHandler);
+                            m_dataChangeQueueHandler = null;
                             break; // queueing is disabled
                         }
 
                         bool queueLastValue = false;
 
-                        if (m_queue == null)
+                        if (m_dataChangeQueueHandler == null)
                         {
-                            m_queue = new MonitoredItemQueue(m_id, QueueOverflowHandler);
+                            m_dataChangeQueueHandler = new DataChangeQueueHandler(Id, IsDurable, m_monitoredItemQueueFactory, QueueOverflowHandler);
                             queueLastValue = true;
                         }
 
-                        m_queue.SetQueueSize(m_queueSize, m_discardOldest, m_diagnosticsMasks);
-                        m_queue.SetSamplingInterval(m_samplingInterval);
+                        m_dataChangeQueueHandler.SetQueueSize(m_queueSize, m_discardOldest, m_diagnosticsMasks);
+                        m_dataChangeQueueHandler.SetSamplingInterval(m_samplingInterval);
 
                         if (queueLastValue && m_lastValue != null)
                         {
-                            m_queue.QueueValue(m_lastValue, m_lastError);
+                            m_dataChangeQueueHandler.QueueValue(m_lastValue, m_lastError);
                         }
                     }
                     else // create event queue.
                     {
-                        if (m_events == null)
-                        {
-                            m_events = new List<EventFieldList>();
-                        }
-
-                        // check if existing queue entries must be discarded;
-                        if (m_events.Count > m_queueSize)
+                        if (m_eventQueueHandler == null)
                         {
-                            int queueSize = (int)m_queueSize;
-
-                            if (m_discardOldest)
-                            {
-                                m_events.RemoveRange(0, m_events.Count - queueSize);
-                            }
-                            else
-                            {
-                                m_events.RemoveRange(queueSize, m_events.Count - queueSize);
-                            }
+                            m_eventQueueHandler = new EventQueueHandler(IsDurable, m_monitoredItemQueueFactory);
                         }
+                        m_eventQueueHandler.SetQueueSize(m_queueSize, m_discardOldest);
```

Comment:
does Setqueue discard the existing items?

Reply:
For Events it does not for better perf. For Data Changes it discards (need to Push that Change, need to fix some Tests before) as overflow Bits need to be set depending on the Queue size .